compiler/src/dotty/tools/dotc/core/quoted/PickledQuotes.scala

```diff
@@ -61,44 +61,117 @@ object PickledQuotes {
-  def unpickleExpr(tasty: PickledQuote, args: PickledArgs)(implicit ctx: Context): Tree = {
+  def unpickleExpr(tasty: PickledQuote, splices: PickledArgs)(implicit ctx: Context): Tree = {
     val tastyBytes = TastyString.unpickle(tasty)
-    val unpickled = unpickle(tastyBytes, args, isType = false)(ctx.addMode(Mode.ReadPositions))
-    /** Force unpickling of the tree, removes the spliced type `@quotedTypeTag type` definitions and dealiases references to `@quotedTypeTag type` */
-    val forceAndCleanArtefacts = new TreeMap {
-      override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree = tree match {
-        case Block(stat :: rest, expr1) if stat.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot) =>
-          assert(rest.forall { case tdef: TypeDef => tdef.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot) })
-          transform(expr1)
-        case tree => super.transform(tree).withType(dealiasTypeTags(tree.tpe))
-      }
-    }
-    forceAndCleanArtefacts.transform(unpickled)
+    val unpickled = unpickle(tastyBytes, splices, isType = false)(ctx.addMode(Mode.ReadPositions))
+    val Inlined(call, Nil, expnasion) = unpickled
+    val inlineCtx = inlineContext(call)
+    val expansion1 = spliceTypes(expnasion, splices)(using inlineCtx)
+    val expansion2 = spliceTerms(expansion1, splices)(using inlineCtx)
+    cpy.Inlined(unpickled)(call, Nil, expansion2)
   }
 
   /** Unpickle the tree contained in the TastyType */
   def unpickleType(tasty: PickledQuote, args: PickledArgs)(implicit ctx: Context): Tree = {
     val tastyBytes = TastyString.unpickle(tasty)
     val unpickled = unpickle(tastyBytes, args, isType = true)(ctx.addMode(Mode.ReadPositions))
-    val tpt = unpickled match {
-      case Block(aliases, tpt) =>
-        // `@quoteTypeTag type` aliases are not required after unpickling.
-        // Type definitions are placeholders for type holes in the pickled quote, at this point
-        // those holes have been filled. As we already dealias al references to them in `dealiasTypeTags`
-        // there is no need to keep their definitions in the tree. As artifacts of quote reification
-        // they also do not have a meaningful position in the source.
-        val aliases1 = aliases.filter(!_.symbol.hasAnnotation(defn.InternalQuoted_QuoteTypeTagAnnot))
-        seq(aliases1, tpt)
-      case tpt => tpt
+    spliceTypes(unpickled, args)
+  }
+
+  /** Replace all term holes with the spliced terms */
+  private def spliceTerms(tree: Tree, splices: PickledArgs)(using Context): Tree = {
+    val evaluateHoles = new TreeMap {
+      override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree = tree match {
+        case Hole(isTerm, idx, args) =>
+          val reifiedArgs = args.map { arg =>
+            if (arg.isTerm) (qctx: scala.quoted.QuoteContext) ?=> new TastyTreeExpr(arg, QuoteContext.scopeId)
+            else new TreeType(arg, QuoteContext.scopeId)
```

Comment:
The syntax `(x:T) ?=> e` is deprecated in favor of `(using x:T) => e`

It's not clear why `qctx` is needed for creation of `TastyTreeExpr`, as from the code, it's not used.